contents/blog/how-we-built-an-app-server.md

```diff
@@ -0,0 +1,304 @@
+
+An app at this point looked like this:
+
+```python
+# exampleplugin/__init__.py
+from posthog.plugins import PluginBaseClass, PosthogEvent, TeamPlugin
+class ExamplePlugin(PluginBaseClass):   
+    def __init__(self, team_plugin_config: TeamPlugin):
+        super().__init__(team_plugin_config)
+        # other per-team init code
+    def process_event(self, event: PosthogEvent):
+        event.properties["hello"] = "world"
+        return event
+    def process_identify(self, event: PosthogEvent):
+        pass
+```
+
+Other features added included a plugin repository, an interface to configure the apps, and a CLI to preconfigure apps for custom installations. Marius built two sample plugins with it: the currency normalizer (convert event properties to a single currency), and the coveted GeoIP app.
+
+It didn’t take long for cracks to appear. First, it’s nearly impossible to manage Python dependencies via multiple `requirements.txt` files. There's no way to tell which dependencies your plugin's dependencies will install, without first installing them. There is [nothing like `pip --dry-run`](https://github.com/pypa/pip/issues/53). You need to install the `pip` package and run an `__init__.py` script inside the package to get its dependency tree.
+
+What's more, all dependencies get installed together. This makes it just a matter of time before a plugin overrides an app dependency (e.g. `django`), and brings everything crashing down. This wasn’t going to work for us, so we needed to rebuild. 
+
+## Rebuilding in JavaScript
+
+After deciding to rebuild, the first attempt was using [PyMiniRacer](https://github.com/sqreen/PyMiniRacer) to generate JavaScript. PyMiniRacer is great for simple functions, but lacked support for async/promises and importing modules like `fetch` because it was raw v8, not Node. It also had a limited standard library. All of which led to the decision it also wasn’t going to work. 
+
+### Another try in JavaScript with gPRC in Node
+
+Marius also tried to develop and subsequently scratched a [gRPC](https://grpc.io/docs/languages/node/basics/) implementation in Node. He was able to call JS code from Python and get a response, but this approach left us with a lot of manual work. It raised questions such as "how many workers should respond to the gRPC calls?" and "how do we make sure we lose no events if the node gRPC server is down?” Apps needed to scale, so this also wasn’t going to work.
+
+### Using Celery
+
+In our main PostHog app, we used [Celery](https://docs.celeryproject.org/) to process events asynchronously. When an event hits `/capture`, our API parses the request and queues the event into a job queue. Marius realized he could build a new server with the [Node port of Celery](https://celery-node.js.org/) and plug that in as another step in the existing pipeline.
```

Comment:
```suggestion
After these false starts, Marius found a solution in something we already used: Celery.

In our main PostHog app, we use [Celery](https://docs.celeryproject.org/) to process events asynchronously. When an event hits `/capture`, our API parses the request and queues the event into a job queue. Marius realized he could build a new server with the [Node port of Celery](https://celery-node.js.org/) and plug that in as another step in the existing pipeline.
```

I'm assuming, possibly incorrectly, we still use Celery? Change back if not!

Reply:
We don't use celery anymore, so I reformatted to past tense.